### PR TITLE
[FLINK-20425][runtime] Change exit code in FatalExitExceptionHandler to be different from JvmShutdownSafeguard

### PR DESCRIPTION
**Pull request by SteNicholas**
## What is the purpose of the change

*`FatalExitExceptionHandler` and `JvmShutdownSafeguard` use the same exit code (-17), making it impossible to distinguish the cause of an unexpected Flink JVM exit based on the exit code alone.  Enum `ExitCode` should be introduced that defines all possible exit codes in Flink. The exit codes are different for `FatalExitExceptionHandler` and `JvmShutdownSafeguard`.*

## Brief change log

  - *Add `ExitCode` enum to indicate the associated process exit code.*
  - *Change the exit code in `FatalExitExceptionHandler` to be different from `JvmShutdownSafeguard`.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-mesos/src/main/java/org/apache/flink/mesos/entrypoint/MesosTaskExecutorRunner.java [hunk 1]
```diff
@@ -33,14 +33,16 @@
...
+
 /**
  * The entry point for running a TaskManager in a Mesos container.
  */
 public class MesosTaskExecutorRunner {
 
 	private static final Logger LOG = LoggerFactory.getLogger(MesosTaskExecutorRunner.class);
 
-	private static final int INIT_ERROR_EXIT_CODE = 31;
+	private static final int INIT_ERROR_EXIT_CODE = RUNNER_INIT.getExitCode();
```
Comment (zentol) [inline flink-mesos/src/main/java/org/apache/flink/mesos/entrypoint/MesosTaskExecutorRunner.java:45]: There's no benefit in keeping these fields.

**Review by zentol — CHANGES_REQUESTED** [review]
I don't quite see the necessity for an enum; all of these could be `static final int`s. Then you could skip the whole `getExitCode()` business.

There are more error codes in the ClusterEntryPoint, AbstractCustomCommandLine, HistoryServer and more that are currently not included.

Ideally these constants are also not defined in the flink-runtime but flink-core so we don't need a dependency on flink-runtime to have access to these codes.

Finally, while consistent error codes are a good thing they should also be documented somewhere.

Simply put, I would suggest to take a step back and either a) only change the exit code of the uncaught fatal exit handler, or properly introduce the constants and do it afterwards.

File: flink-runtime/src/main/java/org/apache/flink/runtime/util/ExitCode.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+public enum ExitCode {
+
+	/** The exit code of the succeeded application process. */
+	SUCCESS(0),
+	/** The exit code for running a TaskExecutor in the container. */
+	RUNNER_INIT(31),
+	/** The exit code of the JVM process killed by the safeguard. */
+	JVM_SHUTDOWN(-17),
+	/** The exit code of the fatal handler for uncaught exceptions. */
+	FATAL_UNCAUGHT(-99);
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/util/ExitCode.java:33]: The comments for all enums aren't ideal.

```suggestion
	/** The process terminated successfully. */
	SUCCESS(0),
	/** The process failed in the initialization of the process. */
	RUNNER_INIT(31),
	/** The process was killed forcefully because the shutdown took too long. */
	JVM_SHUTDOWN(-17),
	/** The process was killed due to an uncaught fatal exception. */
	FATAL_UNCAUGHT(-99);
```

**Comment by zentol** [comment]
I'm closing this PR because of inactivity and it having diverged from the code-base.
